// luma.gl/dist/es5/wip/track-context-creation.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.trackContextCreation = trackContextCreation;

var _assert = _interopRequireDefault(require("../utils/assert"));

// Create a WebGL context

/* global HTMLCanvasElement, WebGLRenderingContext */

/**
 * Installs a spy on Canvas.getContext
 * calls the provided callback with the {context}
 */
function trackContextCreation(_ref) {
  var _ref$onContextCreate = _ref.onContextCreate,
      onContextCreate = _ref$onContextCreate === void 0 ? function () {
    return null;
  } : _ref$onContextCreate,
      _ref$onContextCreated = _ref.onContextCreated,
      onContextCreated = _ref$onContextCreated === void 0 ? function () {} : _ref$onContextCreated;
  (0, _assert.default)(onContextCreate || onContextCreated);

  if (typeof HTMLCanvasElement !== 'undefined') {
    var getContext = HTMLCanvasElement.prototype.getContext;

    HTMLCanvasElement.prototype.getContext = function getContextSpy(type, opts) {
      // Let intercepter create context
      var context;

      if (type === 'webgl') {
        context = onContextCreate({
          canvas: this,
          type: type,
          opts: opts,
          getContext: getContext.bind(this)
        });
      } // If not, create context


      context = context || getContext.call(this, type, opts); // Report it created

      if (context instanceof WebGLRenderingContext) {
        onContextCreated({
          canvas: this,
          context: context,
          type: type,
          opts: opts
        });
      }

      return context;
    };
  }
}
