// luma.gl/dist/es5/webgl-utils/texture-utils.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.cloneTextureFrom = cloneTextureFrom;

var _texture2d = _interopRequireDefault(require("../webgl/texture-2d"));

var _textureCube = _interopRequireDefault(require("../webgl/texture-cube"));

var _texture3d = _interopRequireDefault(require("../webgl/texture-3d"));

var _assert = _interopRequireDefault(require("assert"));

// Clone a new texture object from a reference texture object.
function cloneTextureFrom(refTexture, overrides) {
  (0, _assert.default)(refTexture instanceof _texture2d.default || refTexture instanceof _textureCube.default || refTexture instanceof _texture3d.default);
  var TextureType = refTexture.constructor;
  var gl = refTexture.gl,
      width = refTexture.width,
      height = refTexture.height,
      format = refTexture.format,
      type = refTexture.type,
      dataFormat = refTexture.dataFormat,
      border = refTexture.border,
      mipmaps = refTexture.mipmaps;
  var textureOptions = Object.assign({
    width: width,
    height: height,
    format: format,
    type: type,
    dataFormat: dataFormat,
    border: border,
    mipmaps: mipmaps
  }, overrides); // TODO: move this to `Texture` class as instance method and use this.constructor

  return new TextureType(gl, textureOptions);
}
