// react-map-gl/dist/es5/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "default", {
  enumerable: true,
  get: function get() {
    return _interactiveMap.default;
  }
});
Object.defineProperty(exports, "InteractiveMap", {
  enumerable: true,
  get: function get() {
    return _interactiveMap.default;
  }
});
Object.defineProperty(exports, "StaticMap", {
  enumerable: true,
  get: function get() {
    return _staticMap.default;
  }
});
Object.defineProperty(exports, "BaseControl", {
  enumerable: true,
  get: function get() {
    return _baseControl.default;
  }
});
Object.defineProperty(exports, "Marker", {
  enumerable: true,
  get: function get() {
    return _marker.default;
  }
});
Object.defineProperty(exports, "Popup", {
  enumerable: true,
  get: function get() {
    return _popup.default;
  }
});
Object.defineProperty(exports, "NavigationControl", {
  enumerable: true,
  get: function get() {
    return _navigationControl.default;
  }
});
Object.defineProperty(exports, "CanvasOverlay", {
  enumerable: true,
  get: function get() {
    return _canvasOverlay.default;
  }
});
Object.defineProperty(exports, "HTMLOverlay", {
  enumerable: true,
  get: function get() {
    return _htmlOverlay.default;
  }
});
Object.defineProperty(exports, "SVGOverlay", {
  enumerable: true,
  get: function get() {
    return _svgOverlay.default;
  }
});
Object.defineProperty(exports, "TRANSITION_EVENTS", {
  enumerable: true,
  get: function get() {
    return _transitionManager.TRANSITION_EVENTS;
  }
});
Object.defineProperty(exports, "TransitionInterpolator", {
  enumerable: true,
  get: function get() {
    return _transition.TransitionInterpolator;
  }
});
Object.defineProperty(exports, "LinearInterpolator", {
  enumerable: true,
  get: function get() {
    return _transition.LinearInterpolator;
  }
});
Object.defineProperty(exports, "FlyToInterpolator", {
  enumerable: true,
  get: function get() {
    return _transition.ViewportFlyToInterpolator;
  }
});
exports.experimental = void 0;

var _interactiveMap = _interopRequireDefault(require("./components/interactive-map"));

var _staticMap = _interopRequireDefault(require("./components/static-map"));

var _baseControl = _interopRequireDefault(require("./components/base-control"));

var _marker = _interopRequireDefault(require("./components/marker"));

var _popup = _interopRequireDefault(require("./components/popup"));

var _navigationControl = _interopRequireDefault(require("./components/navigation-control"));

var _canvasOverlay = _interopRequireDefault(require("./overlays/canvas-overlay"));

var _htmlOverlay = _interopRequireDefault(require("./overlays/html-overlay"));

var _svgOverlay = _interopRequireDefault(require("./overlays/svg-overlay"));

var _transitionManager = require("./utils/transition-manager");

var _transition = require("./utils/transition");

var _mapControls = _interopRequireDefault(require("./utils/map-controls"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// Copyright (c) 2015 Uber Technologies, Inc.
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.
// React Map Components
// React Controls
// Overlays
// Utilities
// Experimental Features (May change in minor version bumps, use at your own risk)
var experimental = {
  MapControls: _mapControls.default
};
exports.experimental = experimental;
